cmd/bbolt/main.go

```diff
@@ -1239,7 +1240,7 @@ func (cmd *benchCommand) runWritesWithSource(db *bolt.DB, options *BenchOptions,
 		}); err != nil {
 			return err
 		} else {
-			results.CompletedOps += options.BatchSize
+			results.IncCompletedOps(options.BatchSize)
```

Comment:
Use [AddInt64](https://pkg.go.dev/sync/atomic#AddInt64) instead? 

nit: `IncXXX` seems not good, usually it means incremental by one.

Reply:
I was using the atomic package in the first place just fine, the trouble came with the `time.Duration` which I couldn't figure out how to store/load. I probably made the wrong casting, do you know how to do this? it's just int64 after all..

> nit: IncXXX seems not good, usually it means incremental by one.

happy to change to whatever you like, we can call it `AddCompletedOps` maybe?

Reply:
> happy to change to whatever you like, we can call it AddCompletedOps maybe?

OK


> the trouble came with the time.Duration

Why previous we did not see issue with time.Duration? It may also be updated concurrently previous.

Reply:
sorry, what I was saying is: how would you use the time.Duration with atomics.

```
var d time.Duration
atomic.StoreUint64(&d, time.Now().Sub(time.Second*2))
```

Alternatively, I guess we can use a pointer to a duration instead

Reply:
I know what you are talking about. I tend not to hack the time.Duration. Let's use lock explicitly for simplicity and better readability. 

My concern is why we did not see any race condition for the time.Duration variables. Any confusing is potential issues.

Reply:
The duration wasn't read/changed in the goroutine: https://github.com/etcd-io/bbolt/pull/429/files#diff-11c31bc30063325e6522bd9fe9048b8dd6a3e304adcd54660b72fa5ffbc5ce50R1397-L1406

maybe the `close(finishChan)` acts as an implicit memory barrier? otherwise I do really get it either, the race detector should've pick this up too... I guess there's no test around this that calls the bench method?

Reply:
I am not in the mood of debugging or deep dive in Friday night :). 

Will take a closer look tomorrow. thx.

cc @cenkalti to take a look as well.

Reply:
no worries. I added a test case and with -race enabled it does indeed find this:

```
GOROOT=/usr/local/go #gosetup
GOPATH=/home/tjungblu/go #gosetup
/usr/local/go/bin/go test -c -race -o /tmp/GoLand/___2TestBenchCommand_Run_in_go_etcd_io_bbolt_cmd_bbolt.test go.etcd.io/bbolt/cmd/bbolt #gosetup
/usr/local/go/bin/go tool test2json -t /tmp/GoLand/___2TestBenchCommand_Run_in_go_etcd_io_bbolt_cmd_bbolt.test -test.v -test.paniconexit0 -test.run ^\QTestBenchCommand_Run\E$
=== RUN   TestBenchCommand_Run
=== CONT  TestBenchCommand_Run
    testing.go:1319: race detected during execution of test
--- FAIL: TestBenchCommand_Run (1.14s)
=== RUN   TestBenchCommand_Run/no-args
    main_test.go:274: bench: read: read seq: iter mismatch: expected 1000, got 2000
    --- FAIL: TestBenchCommand_Run/no-args (0.01s)

=== RUN   TestBenchCommand_Run/100k_count
==================
WARNING: DATA RACE
Write at 0x00c0002bd0f0 by goroutine 11:
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWritesWithSource()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1242 +0x99
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWritesSequential()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1201 +0x77
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1177 +0x396
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1096 +0x344
  go.etcd.io/bbolt/cmd/bbolt.(*Main).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:104 +0xb98
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:273 +0x43a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Previous read at 0x00c0002bd0f0 by goroutine 12:
  go.etcd.io/bbolt/cmd/bbolt.checkProgress()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1399 +0x253
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1169 +0x64

Goroutine 11 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:269 +0x244
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 12 (running) created at:
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1169 +0x29a
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1096 +0x344
  go.etcd.io/bbolt/cmd/bbolt.(*Main).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:104 +0xb98
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:273 +0x43a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Write at 0x00c0000fc7f8 by goroutine 11:
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:167 +0x44
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:205 +0xb1
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1101 +0x46d
  go.etcd.io/bbolt/cmd/bbolt.(*Main).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:104 +0xb98
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:273 +0x43a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Previous write at 0x00c0000fc7f8 by goroutine 12:
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:167 +0x44
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:205 +0xb1
  go.etcd.io/bbolt/cmd/bbolt.checkProgress()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1400 +0x12e
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1169 +0x64

Goroutine 11 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:269 +0x244
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 12 (running) created at:
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1169 +0x29a
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1096 +0x344
  go.etcd.io/bbolt/cmd/bbolt.(*Main).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:104 +0xb98
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:273 +0x43a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c0000fc7d8 by goroutine 11:
  bytes.(*Buffer).tryGrowByReslice()
      /usr/local/go/src/bytes/buffer.go:107 +0x52
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:168 +0x18
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:205 +0xb1
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1101 +0x46d
  go.etcd.io/bbolt/cmd/bbolt.(*Main).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:104 +0xb98
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:273 +0x43a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Previous write at 0x00c0000fc7d8 by goroutine 12:
  bytes.(*Buffer).tryGrowByReslice()
      /usr/local/go/src/bytes/buffer.go:108 +0xb3
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:168 +0x18
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:205 +0xb1
  go.etcd.io/bbolt/cmd/bbolt.checkProgress()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1400 +0x12e
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1169 +0x64

Goroutine 11 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:269 +0x244
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 12 (running) created at:
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1169 +0x29a
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1096 +0x344
  go.etcd.io/bbolt/cmd/bbolt.(*Main).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:104 +0xb98
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:273 +0x43a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c0000c2518 by goroutine 11:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:307 +0x0
  bytes.growSlice()
      /usr/local/go/src/bytes/buffer.go:241 +0x14c
  bytes.(*Buffer).grow()
      /usr/local/go/src/bytes/buffer.go:142 +0x2ba
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:170 +0xcd
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:205 +0xb1
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1101 +0x46d
  go.etcd.io/bbolt/cmd/bbolt.(*Main).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:104 +0xb98
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:273 +0x43a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Previous write at 0x00c0000c251a by goroutine 12:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:307 +0x0
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:172 +0x126
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:205 +0xb1
  go.etcd.io/bbolt/cmd/bbolt.checkProgress()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1400 +0x12e
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1169 +0x64

Goroutine 11 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:269 +0x244
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 12 (finished) created at:
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).runWrites()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1169 +0x29a
  go.etcd.io/bbolt/cmd/bbolt.(*benchCommand).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:1096 +0x344
  go.etcd.io/bbolt/cmd/bbolt.(*Main).Run()
      /home/tjungblu/git/bbolt/cmd/bbolt/main.go:104 +0xb98
  go.etcd.io/bbolt/cmd/bbolt_test.TestBenchCommand_Run.func1()
      /home/tjungblu/git/bbolt/cmd/bbolt/main_test.go:273 +0x43a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47
==================
    main_test.go:274: bench: read: read seq: iter mismatch: expected 100000, got 200000
    testing.go:1319: race detected during execution of test
    --- FAIL: TestBenchCommand_Run/100k_count (1.12s)


=== CONT  
    testing.go:1319: race detected during execution of test
FAIL

Process finished with the exit code 1

``` 

I think we should enable the -race detector by default...